From 4b9292b00ce6bcc009098142956656c54f72573e Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Tue, 21 Feb 2017 15:25:41 -0500
Subject: [PATCH] [1/2] Fix a memory leak in jpeg and mpc coder

A leak due to exception handling exist in MPC and JPEG coder.
This could be triggerd by a crafted file.

bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=869791.
origin: https://github.com/ImageMagick/ImageMagick/commit/e5e87c087ed48db886be0ff3aff4041d38218192

(cherry picked from commit e5e87c087ed48db886be0ff3aff4041d38218192)
---
 coders/jpeg.c | 10 ++++++++--
 coders/mpc.c  |  4 +++-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/coders/jpeg.c b/coders/jpeg.c
index 248cef656..f37b1c2d7 100644
--- a/coders/jpeg.c
+++ b/coders/jpeg.c
@@ -1233,7 +1233,10 @@ static Image *ReadJPEGImage(const ImageInfo *image_info,
     }
   if (option != (const char *) NULL)
     if (AcquireImageColormap(image,StringToUnsignedLong(option)) == MagickFalse)
-      ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+      {
+        InheritException(exception,&image->exception);
+        return(DestroyImageList(image));
+      }
   if ((jpeg_info.output_components == 1) && (jpeg_info.quantize_colors == 0))
     {
       size_t
@@ -1241,7 +1244,10 @@ static Image *ReadJPEGImage(const ImageInfo *image_info,
 
       colors=(size_t) GetQuantumRange(image->depth)+1;
       if (AcquireImageColormap(image,colors) == MagickFalse)
-        ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+        {
+          InheritException(exception,&image->exception);
+          return(DestroyImageList(image));
+        }
     }
   if (image->debug != MagickFalse)
     {
diff --git a/coders/mpc.c b/coders/mpc.c
index 72fb69a08..64e92ed31 100644
--- a/coders/mpc.c
+++ b/coders/mpc.c
@@ -842,7 +842,9 @@ static Image *ReadMPCImage(const ImageInfo *image_info,ExceptionInfo *exception)
         /*
           Create image colormap.
         */
-        if (AcquireImageColormap(image,image->colors) == MagickFalse)
+        image->colormap=(PixelInfo *) AcquireQuantumMemory(image->colors+1,
+          sizeof(*image->colormap));
+        if (image->colormap == (PixelInfo *) NULL)
           ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
         if (image->colors != 0)
           {
